File: features/task/src/main/java/com/nivasafinance/features/task/exception/TaskValidationException.java
package com.nivasafinance.features.task.exception;

import com.nivasafinance.common.exception.ExceptionUtils;
import com.nivasafinance.common.exception.ValidationException;
import org.springframework.context.MessageSource;

import java.io.Serial;

public class TaskValidationException extends ValidationException {

    @Serial
    private static final long serialVersionUID = 1234567890123456791L;

    private TaskValidationException(String message) {
        super(message);
    }

    public static TaskValidationException requestRequired(MessageSource messageSource) {
        return new TaskValidationException(
            ExceptionUtils.createLocalizedMessage("error.task.request.required", null, messageSource)
        );
    }

    public static TaskValidationException taskIdRequired(MessageSource messageSource) {
        return new TaskValidationException(
            ExceptionUtils.createLocalizedMessage("error.task.id.required", null, messageSource)
        );
    }

    public static TaskValidationException taskConfigKeyRequired(MessageSource messageSource) {
        return new TaskValidationException(
            ExceptionUtils.createLocalizedMessage("error.task.config.key.required", null, messageSource)
        );
    }

    public static TaskValidationException assignmentRequired(MessageSource messageSource) {
        return new TaskValidationException(
            ExceptionUtils.createLocalizedMessage("error.task.assignment.required", null, messageSource)
        );
    }

    public static TaskValidationException outcomeRequired(MessageSource messageSource) {
        return new TaskValidationException(
            ExceptionUtils.createLocalizedMessage("error.task.outcome.required", null, messageSource)
        );
    }

    public static TaskValidationException newAssignmentRequired(MessageSource messageSource) {
        return new TaskValidationException(
            ExceptionUtils.createLocalizedMessage("error.task.new.assignment.required", null, messageSource)
        );
    }

    public static TaskValidationException dueDateRequired(MessageSource messageSource) {
        return new TaskValidationException(
            ExceptionUtils.createLocalizedMessage("error.task.due.date.required", null, messageSource)
        );
    }

    public static TaskValidationException invalidOutcome(String outcome, String taskConfigKey, MessageSource messageSource) {
        return new TaskValidationException(
            ExceptionUtils.createLocalizedMessage("error.task.outcome.invalid", 
                new Object[]{outcome, taskConfigKey}, messageSource)
        );
    }

    public static TaskValidationException userCannotCreateTask(String username, String taskConfigKey, MessageSource messageSource) {
        return new TaskValidationException(
            ExceptionUtils.createLocalizedMessage("error.task.user.cannot.create", 
                new Object[]{username, taskConfigKey}, messageSource)
        );
    }

    public static TaskValidationException userCannotCompleteTask(String username, String taskConfigKey, MessageSource messageSource) {
        return new TaskValidationException(
            ExceptionUtils.createLocalizedMessage("error.task.user.cannot.complete", 
                new Object[]{username, taskConfigKey}, messageSource)
        );
    }

    public static TaskValidationException assignedUserDoesNotHaveRequiredRole(String assignedTo, String taskConfigKey, MessageSource messageSource) {
        return new TaskValidationException(
            ExceptionUtils.createLocalizedMessage("error.task.assigned.user.invalid.role", 
                new Object[]{assignedTo, taskConfigKey}, messageSource)
        );
    }

    public static TaskValidationException entityTypeRequired(MessageSource messageSource) {
        return new TaskValidationException(
            ExceptionUtils.createLocalizedMessage("error.task.entity.type.required", null, messageSource)
        );
    }

    public static TaskValidationException entityIdRequired(MessageSource messageSource) {
        return new TaskValidationException(
            ExceptionUtils.createLocalizedMessage("error.task.entity.id.required", null, messageSource)
        );
    }

    public static TaskValidationException entityDetailsRequired(MessageSource messageSource) {
        return new TaskValidationException(
            ExceptionUtils.createLocalizedMessage("error.task.entity.details.required", null, messageSource)
        );
    }
}

File: features/tasks/src/main/java/com/nivasafinance/features/tasks/exception/TaskValidationException.kt
package com.nivasafinance.features.tasks.exception

import exception.ExceptionUtils
import exception.ValidationException
import org.springframework.context.MessageSource
import java.util.*

open class TaskValidationException(
    messageKey: String,
    args: Array<Any>? = null,
    messageSource: MessageSource
) : ValidationException(
    ExceptionUtils.createLocalizedMessage(messageKey, args, messageSource)
)

class TaskDueDatePastException(messageSource: MessageSource) : TaskValidationException(
    "error.task.due.date.past",
    null,
    messageSource
)

class TaskAssignmentRequiredException(
    status: String,
    messageSource: MessageSource
) : TaskValidationException(
    "error.task.assignment.required",
    arrayOf(status),
    messageSource
)

class TaskCannotUpdateCompletedException(
    taskId: UUID,
    messageSource: MessageSource
) : TaskValidationException(
    "error.task.cannot.update.completed",
    arrayOf(taskId.toString()),
    messageSource
)

class TaskCannotDeleteCompletedException(
    taskId: UUID,
    messageSource: MessageSource
) : TaskValidationException(
    "error.task.cannot.delete.completed",
    arrayOf(taskId.toString()),
    messageSource
)

class TaskInvalidStatusTransitionException(
    currentStatus: String,
    newStatus: String,
    messageSource: MessageSource
) : TaskValidationException(
    "error.task.invalid.status.transition",
    arrayOf(currentStatus, newStatus),
    messageSource
)
